lib/blas_3d.cu

```diff
@@ -0,0 +1,251 @@
+
+    public:
+      axpby3D(ColorSpinorField &x, ColorSpinorField &y, const std::vector<double> &a, const std::vector<double> &b) :
+        TunableKernel2D(x, x.SiteSubset()), x(x), y(y), a(a), b(b)
+      {
+        apply(device::get_default_stream());
+      }
+
+      void apply(const qudaStream_t &stream) override
+      {
+        TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+        launch<axpby3d>(tp, stream, axpby3dArg<Float, nColor>(a, x, b, y));
+      }
+
+      void preTune() override { y.backup(); }
+      void postTune() override { y.restore(); }
+      long long flops() const override { return 6 * x.Volume() * x.Nspin() * x.Ncolor(); }
+      long long bytes() const override { return x.Bytes() + 2 * y.Bytes(); }
+    };
+
+    void axpby(std::vector<double> &a, ColorSpinorField &x, std::vector<double> &b, ColorSpinorField &y)
+    {
+      checkPrecision(x, y);
+
+      // Check spins
+      if (x.Nspin() != y.Nspin()) errorQuda("Unexpected number of spins x=%d y=%d", x.Nspin(), y.Nspin());
+
+      // Check colors
+      if (x.Ncolor() != y.Ncolor()) errorQuda("Unexpected number of colors x=%d y=%d", x.Ncolor(), y.Ncolor());
+
+      // Check coefficients
+      if (a.size() != b.size() && a.size() != (unsigned int)x.X()[3])
+        errorQuda("Unexpected coeff array sizes a=%lu b=%lu, x[3]=%d", a.size(), b.size(), x.X()[3]);
+
+      // We must give a Lattice field as the first argument
+      instantiate<axpby3D>(x, y, a, b);
+    }
+
+    void ax(std::vector<double> &a, ColorSpinorField &x)
+    {
+      std::vector<double> zeros(a.size(), 0.0);
+      axpby(a, x, zeros, x);
+    }
+
+    template <typename Float, int nColor> class caxpby3D : TunableKernel2D
+    {
+      ColorSpinorField &x;
+      ColorSpinorField &y;
+      const std::vector<Complex> &a;
+      const std::vector<Complex> &b;
+      unsigned int minThreads() const override { return x.VolumeCB(); }
+
+    public:
+      caxpby3D(ColorSpinorField &x, ColorSpinorField &y, const std::vector<Complex> &a, const std::vector<Complex> &b) :
+        TunableKernel2D(x, x.SiteSubset()), x(x), y(y), a(a), b(b)
+      {
+        apply(device::get_default_stream());
+      }
+
+      void apply(const qudaStream_t &stream) override
+      {
+        TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+        launch<caxpby3d>(tp, stream, caxpby3dArg<Float, nColor>(a, x, b, y));
+      }
+
+      void preTune() override { y.backup(); }
+      void postTune() override { y.restore(); }
+      long long flops() const override { return 14 * x.Volume() * x.Nspin() * x.Ncolor(); }
+      long long bytes() const override { return x.Bytes() + 2 * y.Bytes(); }
+    };
+
+    void caxpby(std::vector<Complex> &a, ColorSpinorField &x, std::vector<Complex> &b, ColorSpinorField &y)
+    {
+      checkPrecision(x, y);
+
+      // Check spins
+      if (x.Nspin() != y.Nspin()) errorQuda("Unexpected number of spins x=%d y=%d", x.Nspin(), y.Nspin());
```

Comment:
Maybe out of the scope of this PR, but either here or in the future (I can create an issue, let me know) we should create a set of routines `checkSpin`, `checkColor`, `checkSpinColor` that acts similarly to `checkPrecision`, etc.

Reply:
I just went and did it dfd36e754